source/simulator/stokes_matrix_free.cc

```diff
@@ -1697,6 +1764,114 @@ namespace aspect
+              const unsigned int n_faces_interior = stokes_matrix.get_matrix_free()->n_inner_face_batches();
+
+              active_cell_data.free_surface_boundary_indicators =
+                sim.mesh_deformation->get_free_surface_boundary_indicators();
+
+              MaterialModel::MaterialModelInputs<dim> face_material_inputs(n_face_q_points, sim.introspection.n_compositional_fields);
+              MaterialModel::MaterialModelOutputs<dim> face_material_outputs(n_face_q_points, sim.introspection.n_compositional_fields);
+
+              active_cell_data.free_surface_stabilization_term_table.reinit(n_faces_boundary, n_face_q_points);
+
+              for (unsigned int face=n_faces_interior; face<n_faces_boundary + n_faces_interior; ++face)
+                {
+                  const unsigned int n_components_filled = stokes_matrix.get_matrix_free()->n_active_entries_per_face_batch(face);
+
+                  for (unsigned int i=0; i<n_components_filled; ++i)
+                    {
+                      // the face iterator can be accessed: pair.first->face(pair.second);
+                      // that means pair.second() is the face number.
+                      const auto cell_face_pair = stokes_matrix.get_matrix_free()->get_face_iterator(face, i, true);
+
+                      typename DoFHandler<dim>::active_cell_iterator matrix_free_cell =
+                        cell_face_pair.first;
+                      typename DoFHandler<dim>::active_cell_iterator simulator_cell(&(sim.triangulation),
+                                                                                    matrix_free_cell->level(),
+                                                                                    matrix_free_cell->index(),
+                                                                                    &(sim.dof_handler));
+
+                      const types::boundary_id boundary_indicator = stokes_matrix.get_matrix_free()->get_boundary_id(face);
+                      AssertDimension(boundary_indicator, simulator_cell->face(cell_face_pair.second)->boundary_id());
+
+                      // only apply on free surface faces
+                      if (active_cell_data.free_surface_boundary_indicators.find(boundary_indicator)
+                          == active_cell_data.free_surface_boundary_indicators.end())
+                        continue;
+
+                      fe_face_values.reinit(simulator_cell, cell_face_pair.second);
+
+                      face_material_inputs.reinit(fe_face_values, simulator_cell, sim.introspection, sim.solution);
+
+                      sim.compute_material_model_input_values(sim.solution, fe_face_values,
+                                                              simulator_cell, false, face_material_inputs);
+                      sim.material_model->evaluate(face_material_inputs, face_material_outputs);
```

Comment:
Do we do averaging of the surface terms in matrix-based?

Reply:
No we don't, I will use harmonic average in the matrix-based one.